Kudu.Core/SiteExtensions/SiteExtensionStatus.cs

```diff
@@ -104,6 +117,74 @@ public bool IsTerminalStatus()
                 || string.Equals(Constants.SiteExtensionProvisioningStateCanceled, ProvisioningState, StringComparison.OrdinalIgnoreCase);
         }
 
+        /// <summary>
+        /// <para>Loop thru all site extension settings files under 'siteExtensionStatusRoot', check if there is any successful installation</para>
+        /// <para>if not install to webroot, will trigger restart; if install to webroot and with applicationHost.xdt file, trigger restart.</para>
+        /// </summary>
+        /// <param name="siteExtensionStatusRoot">should be $ROOT\site\siteextensions</param>
+        /// <param name="siteExtensionRoot">should be $ROOT\SiteExtensions</param>
+        public static bool IsAnyInstallationRequireRestart(string siteExtensionStatusRoot, string siteExtensionRoot, ITracer tracer, IAnalytics analytics)
+        {
+            try
+            {
+                string[] packageDirs = FileSystemHelpers.GetDirectories(siteExtensionStatusRoot);
+                foreach (var dir in packageDirs)
+                {
+                    try
+                    {
+                        DirectoryInfo dirInfo = new DirectoryInfo(dir);
+                        string[] settingFile = FileSystemHelpers.GetFiles(dir, _statusSettingsFileName);
+                        foreach (var file in settingFile)
+                        {
+                            var statusSettings = new SiteExtensionStatus(siteExtensionStatusRoot, dirInfo.Name, tracer);
+                            if (IsSiteExtensionRequireRestart(statusSettings, siteExtensionRoot))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                    catch(Exception ex)
+                    {
+                        analytics.UnexpectedException(ex, trace: false);
+                        tracer.TraceError(ex);
+                        tracer.TraceError("Failed to query {0} under {1}, continus to check others ...", _statusSettingsFileName, dir);
```

Comment:
not combining the two "tracer.TraceError".

when perform "tracer.TraceError(ex)", "text" will be "ex.Message"
when perform "tracer.TraceError(message, param)", "text" will be given message
there is no interface for "tracer.TraceError(ex, message)"

if it is necessary, we can create "tracer.TraceError(ex, message, params)"
and assign combine given message and ex.Message into "text" attribute

```C#
        public static void TraceError(this ITracer tracer, Exception ex)
        {
            var attribs = new Dictionary<string, string>
            {
                { "type", "error" },
                { "text", ex.Message },
                { "stackTrace", ex.StackTrace ?? String.Empty }
            };

            if (ex.InnerException != null)
            {
                attribs["innerText"] = ex.InnerException.Message;
                attribs["innerStackTrace"] = ex.InnerException.StackTrace ?? String.Empty;
            }

            tracer.Trace("Error occurred", attribs);
        }

        public static void TraceError(this ITracer tracer, string message)
        {
            tracer.Trace("Error occurred", new Dictionary<string, string>
            {
                { "type", "error" },
                { "text", message }
            });
        }
```